yt/frontends/halo_catalog/data_structures.py

```diff
@@ -146,3 +177,363 @@ def _is_valid(self, *args, **kwargs):
+
+    def _calculate_particle_index_starts(self):
+        """
+        Create a dict of halo id offsets for each file.
+        """
+        particle_count = defaultdict(int)
+        offset_count = 0
+        for data_file in self.data_files:
+            data_file.index_start = dict(
+                [(ptype, particle_count[ptype]) for ptype in data_file.total_particles]
+            )
+            data_file.offset_start = offset_count
+            for ptype in data_file.total_particles:
+                particle_count[ptype] += data_file.total_particles[ptype]
+            offset_count += getattr(data_file, "total_offset", 0)
+
+        self._halo_index_start = dict(
+            [
+                (
+                    ptype,
+                    np.array(
+                        [data_file.index_start[ptype] for data_file in self.data_files]
+                    ),
+                )
+                for ptype in self.ds.particle_types_raw
+            ]
+        )
+
+    def _create_halo_id_table(self):
+        pass
+
+    def _detect_output_fields(self):
+        field_list = []
+        scalar_field_list = []
+        units = {}
+        pc = dict(
+            [
+                (ptype, sum([d.total_particles[ptype] for d in self.data_files]))
+                for ptype in self.ds.particle_types_raw
+            ]
+        )
+        found_fields = dict([(ptype, False) for ptype, pnum in pc.items() if pnum > 0])
+        has_ids = False
+
+        for data_file in self.data_files:
+            fl, sl, idl, _units = self.io._identify_fields(data_file)
+            units.update(_units)
+            field_list.extend([f for f in fl if f not in field_list])
+            scalar_field_list.extend([f for f in sl if f not in scalar_field_list])
+            for ptype in found_fields:
+                found_fields[ptype] |= data_file.total_particles[ptype]
+            has_ids |= len(idl) > 0
+            if all(found_fields.values()) and has_ids:
+                break
+
+        self.field_list = field_list
+        self.scalar_field_list = scalar_field_list
+        ds = self.dataset
+        ds.scalar_field_list = scalar_field_list
+        ds.particle_types = tuple(set(pt for pt, ds in field_list))
+        ds.field_units.update(units)
+        ds.particle_types_raw = ds.particle_types
+
+    def _get_halo_file_indices(self, ptype, identifiers):
+        return np.digitize(identifiers, self._halo_index_start[ptype], right=False) - 1
+
+    def _get_halo_scalar_index(self, ptype, identifier):
+        i_scalar = self._get_halo_file_indices(ptype, [identifier])[0]
+        scalar_index = identifier - self._halo_index_start[ptype][i_scalar]
+        return scalar_index
+
+    def _get_halo_values(self, ptype, identifiers, fields, f=None):
+        """
+        Get field values for halo data containers.
+        """
+
+        # if a file is already open, don't open it again
+        filename = None if f is None else f.filename
```

Comment:
```suggestion
        try:
            filename = f.filename
        except AttributeError:
            filename = None
```

Reply:
Ok well it's not crucial here since the function isn't user-facing, but for the record, I find that try-except blocks are an extremely useful pattern for handling user data. Most of the bugs I've been dealing with recently are solved by changing ifs into tries. But then again I didn't realise performance was an issue here, so this is fine :)